### PR TITLE
GSoC butteraugli

### PR DESCRIPTION
**Pull request by sn99**
What's done: 
- bindings
- few tests
- part conversion to pure rust

What's left:
- Own implementation of `Image` Struct
- Add changes made by libjxl to butteraulgi
- Add benchmarks

Related blog entries:
- [Butteraugli: Google vs libjxl](https://sn99.github.io/2021/06/07/butteraugli-google-vs-libjxl/)
- [First week of GSoC](https://sn99.github.io/2021/06/15/first-week-of-gsoc/)
- [Ending: GSoC](https://sn99.github.io/2021/08/28/ending-gsoc/)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,7 +1,10 @@
 [package]
-name = "butteraugli"
+name = "butteraugli-rs"
```
Comment (lu-zero) [inline Cargo.toml]: the crate name remains `butteraugli`. Please fix it.

File: butteraugli-sys/.cargo/config [hunk 2]
```diff
@@ -0,0 +1,2 @@
+[build]
+rustflags = ["-C", "link-args=-lpng -ljpeg"]
```
Comment (lu-zero) [inline butteraugli-sys/.cargo/config]: You should use [systemdeps](https://crates.io/crates/system-deps).

**Review by lu-zero — CHANGES_REQUESTED** [review]
Does the code at least build?

File: src/image.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+use std::mem;
+
+pub struct ImageF {
+    pub xsize_: usize,
+    pub ysize_: usize,
```
Comment (lu-zero) [inline src/image.rs]: Remove the `_` at the end

File: src/butteraugli.rs [hunk 4]
```diff
@@ -977,3 +977,45 @@ pub fn new_srgb_to_linear_table() -> [f64; 256] {
...
+    let mut weight = 0.0;
+
+    for j in minx..=maxx {
+        weight += kernel[(j - x + offset) as usize];
+    }
+
+    // Interpolate linearly between the no-border scaling and border scaling.
+    weight = (1.0 - border_ratio) * weight + border_ratio * weight_no_border;
+    let scale = 1.0 / weight;
+    for y in 0..in_.ysize() {
```
Comment (lu-zero) [inline src/butteraugli.rs:1014]: this would be the `ysize` field.

File: src/butteraugli.rs [hunk 5]
```diff
@@ -977,3 +977,45 @@ pub fn new_srgb_to_linear_table() -> [f64; 256] {
...
+        let row_in = in_.Row(y);
```
Comment (lu-zero) [inline src/butteraugli.rs:1015]: Look at how v_frame has `row()` and `rows_iter()` per-plane.

File: src/butteraugli.rs [hunk 6]
```diff
@@ -977,3 +977,45 @@ pub fn new_srgb_to_linear_table() -> [f64; 256] {
...
+        let mut sum = 0.0;
+        for j in minx..=maxx {
+            sum += row_in[j] * kernel[j - x + offset];
+        }
+        row_out[y] = sum * scale;
```
Comment (lu-zero) [inline src/butteraugli.rs:1020]: row_out cannot be a f64, it looks like it is a slice

**Comment by sn99** [comment]
Forgot to comment our parts of Image, some functions are still a placeholder, bindings build fine

File: .cargo/config [hunk 7]
```diff
@@ -0,0 +1,2 @@
+[build]
+rustflags = ["-C", "link-args=-lpng -ljpeg"]
```
Comment (lu-zero) [inline .cargo/config]: This is not the way to do it, I told you to use [system-deps](https://crates.io/crates/system-deps) ...

Comment (sn99) [reply .cargo/config]: I unresolved the previous one already, I was resolving another one and misclicked this one

**Pull request closed without merging**